model/generate.go

```diff
@@ -161,8 +161,17 @@ func (g *GeneratedProject) Save(ctx context.Context, p *Project, pp *ParserProje
 	if err := updateParserProject(v, pp); err != nil {
 		return errors.WithStack(err)
 	}
+	// Now that we've already updated the project, mark tasks as generated
+	// to verify we don't hit re-defining errors with other jobs while saving new builds/tasks.
+	if err := task.MarkGeneratedTasks(t.Id, true); err != nil {
+		return errors.Wrap(err, "error marking tasks as generated")
+	}
 
 	if err := g.saveNewBuildsAndTasks(ctx, v, p, t); err != nil {
+		// Because we were unable to save new builds and tasks, release the generator lock.
+		if err2 := task.MarkGeneratedTasks(t.Id, false); err2 != nil {
```

Comment:
I think this trades one problem for another, but I don't have a good sense of which problem would be more common. I think the problem is that if there is an app crash or panic during `saveNewBuildsAndTasks`, then the task will be marked as having generated tasks despite not having generated. To be correct a lock like this would also need a lease, which is probably more complex than we want. (See https://martin.kleppmann.com/2016/02/08/how-to-do-distributed-locking.html for more detail about this.)

Reply:
Ah this is a good point. I think that problem would be less common, though still possible. I think the way to handle this instead then is to still retry on config errors, but we should retry less (either just for config errors or in general). I'll write a new thing.